lib/verify-jwt.js

```diff
@@ -1,31 +1,37 @@
 // @flow
+
 const jwt = require('jsonwebtoken')
-module.exports = async (token /* : string */, jwks /* : JwksClient */) => {
-  try {
-    if (token.includes('Bearer ')) {
-      token = token.split(' ')[1]
-    }
-    const { header } = jwt.decode(token, { complete: true })
-    const publicKey = await getPublicKey(jwks, header.kid)
 
-    const decoded = jwt.verify(token, publicKey, {
-      algorithms: ['RS256'],
-      issuer: jwks.options.jwksUri.split('/.well-known')[0], // ISS
-    })
-    return decoded
-  } catch (err) {
-    throw new Error('The supplied JWT was invalid.')
+module.exports = async (token /* : string */, jwks /* : JwksClient */) => {
+  if (token.includes('Bearer ')) {
+    token = token.split(' ')[1]
   }
-}
 
-const getPublicKey = (
-  jwks /* : JwksClient */,
-  kid /* : string */,
-) /* : Promise<string> */ => {
-  return new Promise((resolve, reject) => {
-    jwks.getSigningKey(kid, (err, key) => {
-      if (err) return reject(err)
-      return resolve(key.publicKey || key.rsaPublicKey)
-    })
+  const decoded = await new Promise((resolve, reject) => {
+    jwt.verify(
+      token,
+      (header, callback) => {
+        jwks.getSigningKey(header.kid, (error, key) => {
+          if (error) {
+            callback(error)
+          } else {
+            callback(null, key.publicKey || key.rsaPublicKey)
+          }
+        })
+      },
+      {
+        algorithms: ['RS256'],
+        issuer: jwks.options.jwksUri.split('/.well-known')[0], // ISS
+      },
+      (error, decoded) => {
+        if (error) {
+          reject(error)
+        } else {
+          resolve(decoded)
+        }
+      },
+    )
   })
+
+  return decoded
 }
```

Comment:
I get why you did this, but it seems a bunch more convoluted?

Reply:
Yeah, code is not a pretty, but only decoding the JWT once now.

Reply:
Ill clean it up slightly :)

Reply:
Alrighty